File: solutions/r/beer-song/beer-song.R
# Get the entire beer song for a given range of numbers of bottles.
lyrics <- function(first, last) {
    (verse(first:last) |> paste(collapse = "\n"))
}

# Get a single verse of the beer song
verse <- Vectorize(function(number) {
    verse_last <- paste0(
        "No more bottles of beer on the wall, no more bottles of beer.\n",
        "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
    )
    verse_template <- function(count1, count2, one = "one") {
        paste0(
            count1, " of beer on the wall, ",
            count1, " of beer.\n",
            "Take ", one, " down and pass it around, ",
            count2, " of beer on the wall.\n"
        )
    }

    if (number == 0) {
        verse_last
    } else if (number == 1) {
        verse_template("1 bottle", "no more bottles", "it")
    } else if (number == 2) {
        verse_template("2 bottles", "1 bottle")
    } else {
        verse_template(paste(number, "bottles"), paste(number - 1, "bottles"))
    }
})

File: solutions/elixir/beer-song/lib/beer_song.ex
defmodule BeerSong do
  @moduledoc false

  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t()
  def verse(0) do
    """
    No more bottles of beer on the wall, no more bottles of beer.
    Go to the store and buy some more, 99 bottles of beer on the wall.
    """
  end

  def verse(1), do: verse_template("1 bottle", "no more bottles", "it")
  def verse(2), do: verse_template("2 bottles", "1 bottle")
  def verse(number), do: verse_template("#{number} bottles", "#{number - 1} bottles")

  @doc false
  @spec verse_template(String.t(), String.t(), String.t()) :: String.t()
  defp verse_template(count1, count2, one \\ "one") do
    """
    #{count1} of beer on the wall, #{count1} of beer.
    Take #{one} down and pass it around, #{count2} of beer on the wall.
    """
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range \\ 99..0), do: Enum.map_join(range, "\n", &verse/1)
end
